pkg/kubelet/kubelet_pods.go

```diff
@@ -472,11 +472,17 @@ func (kl *Kubelet) GenerateRunContainerOptions(pod *v1.Pod, container *v1.Contai
 			glog.Errorf("Error on creating %q: %v", p, err)
 		} else {
 			opts.PodContainerDir = p
+			if err := kl.chownDirForRemappedIDs(p); err != nil {
+				return nil, nil, fmt.Errorf("error while chowning %s", p)
+			}
+			if err := kl.chownDirForRemappedIDs(filepath.Join(p, "..")); err != nil {
```

Comment:
cc @kubernetes/sig-storage-pr-reviews @kubernetes/sig-auth-pr-reviews for eyes on path traversal possibilities

Reply:
@liggitt, thanks for heads-up.

A lot of existing volume plugins use [FSgroup](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2665) to make  data on a volume (local or remote) available to a pod. Kubernetes does `chown` and `chmod`  with the GID specified in `FSGroup`.

Let's say a pod has FSGroup = 500 and uses a volume. Kubernetes chmods and chowns the files on the volume to be accessible by group 500 and run a container with supplementary group 500. How will this work with user namespace? What supplementary group will have a process in the container? Will it be able to access **all** data on the volume?


As a separate item, we should consider how to remove chown/chmod done for FSGroup in Kubernetes and use user namespaces instead, as it's time consuming.

Reply:
@jsafrane fsGroup will not work as intended. On the host fsgroup will get remapped as per usernamespace configuration and remapped ID will not match with group ID on the remote volume. This is because of lack of support in runtimes. However, we can aim to get this working before graduating feature to beta.

Reply:
If I read this correctly, that means all pods that don't use `PodSpec.HostUser` can't use any volumes with `FSGroup`? That makes most volumes unusable in multi-tenant environment.

Reply:
@vikaschoudhary16 can you explain how you think this limitation would be addressed?  it's not immediately clear to me at this time, but it seems worth understanding at a high-level now.

Reply:
it would help if we can flesh out https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/node-usernamespace-remapping.md#remote-volumes to know what the path is for actually graduating this feature.  right now, in alpha form, without this it would be very limiting.

Reply:
ascending to the parent dir via `..` reminds me of the issue https://github.com/kubernetes/kubernetes/issues/57095 ... is there risk of hitting the wrong location by joining ".."?

Reply:
@derekwaynecarr This behavior is runtime-dependent. As per current behaviour in runtimes for example crio and docker, where FsGroup mentioned in pod spec is translated by runtime, if userns is enabled, as the actual gid on the volume files. Ex: x->f(x) is mapping, f(x) will be gid, where 'x' is FsGroup. This would work fine if the volume is shared by the pods who have same userns settings i.e either enabled at all the pods(nod) with same userns configuration or all using host usernamespace.

Problem will be there if there is a mismatch in user-namespace configuration.
May be if runtimes interpret FsGroup ID as a translated one i.e actual GID on host, in that case it might work for all the cases. This will require some handling to be added at runtimes eg: crio , docker (any runtime).
Will it be ok to go alpha with this userns configuration matching requirement limitation for FsGroup and as a roadmap, before beta, target to get the handling changed/added at supported runtimes?

/cc @mrunalp @giuseppe

Reply:
@liggitt nice catch, will fix this. Thanks!